src/coreclr/jit/emitarm64.cpp

```diff
@@ -10253,7 +10253,7 @@ unsigned emitter::emitOutputCall(insGroup* ig, BYTE* dst, instrDesc* id, code_t
 /*static*/ unsigned emitter::emitOutput_Instr(BYTE* dst, code_t code)
 {
     assert(sizeof(code_t) == 4);
-    *((code_t*)dst) = code;
+    *((code_t*)(dst + writeableOffset)) = code;
```

Comment:
If I understand correctly we keeep `dst` name for `dstRE` and name all RW accesses as  `dstRW`?

I would prefer to see a named variable here, if I understand correctly it is
```
unsigned emitter::emitOutput_Instr(BYTE* dstRE, code_t code)
BYTE* dstRW = dst + writeableOffsetl
```

Reply:
That would require quite a lot of changes over the JIT code base as all code addresses in JIT are read-execute and the RW are used only at places where it is needed. So everywhere that dst is used now, it would become dstRE. That seemed like a noise change to me that would obscure the real changes.

> what would happen if a jit developer uses a wrong address for a write (dstRW instead of dstRE, or dstRE instead of dstRW)

If you attempt to write using dstRE address when jitting, the JIT would crash as it would not be able to write there. If you use the dstRW address instead of dstRE, it would be fine unless you use it for code relative address computation.
I've tried to use the dstRW only locally so that it would be hard to make such a mistake. The only exception is the NOP generating code that calls itself recursively and it would have to update the RE pointers to match the RW all the time.
In fact, I'd rather remove the dstRW usage at other places and change it to (dst + writeableOffset).

> will we see the issue on the platforms without W^X protection?

There will be no such platforms (except of crossgen usage of the JIT)

> Can we catch them at compilation time?

We would have to change all the `BYTE* dst` to `const BYTE* dst`, but it would still require the JIT developer to be disciplined enough to not to cast away the constness anywhere.

Frankly, my view on further updates to JIT were that everything would be done as before with the only exception being places where writes to the executable memory are added. At those places, the writeableOffset would need to be applied. But a bug in such a place would be easy and quick to discover - the JIT would just crash with access violation.

Reply:
I've actually decided to create and use dstRW at all places. The one above was actually the only one that was not done that way.